plugins/AddLeptonJetRelatedVariables.cc

```diff
@@ -0,0 +1,282 @@
+  iEvent.getByToken(vertexes_, PVs);
+  reco::VertexRef PV(PVs.id());
+  reco::VertexRefProd PVRefProd(PVs);
+
+  //Output
+  std::vector<double > ptratio,ptrel,nchargeddaughers,btagcsv;
+  for( reco::CandidateView::const_iterator icand = leptons->begin(); icand != leptons->end(); ++ icand){
+
+    //Initialise loop variables
+    float dR = 9999;
+    reco::Candidate::LorentzVector jet, mu;
+    float bjet(-999.);
+    int nchdaugthers(0);    
+    double jecL1L2L3Res = 1.;
+    double jecL1 = 1.;
+    
+    for (reco::JetTagCollection::const_iterator tagI = bTags.begin(); 
+	 tagI != bTags.end(); ++tagI) {
+
+      // for each muon with the lepton 
+      if(deltaR(*(tagI->first), *icand) > dR) continue;  
+      dR = deltaR(*(tagI->first), *icand);
+      
+      mu  = icand->p4(); 
+      jet = tagI->first->p4();
+
+      jecL1L2L3Res = correctorL1L2L3Res->correction(*(tagI->first));
+      jecL1 = correctorL1->correction(*(tagI->first));
+
+      // Get b-jet info
+      bjet = tagI->second;
+      unsigned int ic=0;
+      for (PFCollection::const_iterator iP = pfCandidates->begin(); iP !=pfCandidates->end(); ++iP){
+	ic++;
+	const reco::PFCandidate *pfcand = iP->get();
+	if (pfcand->charge()==0) continue;
+	if (deltaR(*tagI->first,*pfcand) > 0.4) continue;	
+	if (!pfcand->bestTrack()) continue;
+	if (!pfcand->bestTrack()->quality(reco::Track::highPurity)) continue;
+	if (pfcand->bestTrack()->pt()<1.) continue;
+	if (pfcand->bestTrack()->hitPattern().numberOfValidHits()<8) continue;
+	if (pfcand->bestTrack()->hitPattern().numberOfValidPixelHits()<2) continue;
+	if (pfcand->bestTrack()->normalizedChi2()>=5) continue;
+	
+	PV = reco::VertexRef(PVs, 0);
+	math::XYZPoint PVpos = PV->position();
+
+	if (std::fabs(pfcand->bestTrack()->dxy(PVpos)) > 0.2) continue;
+	if (std::fabs(pfcand->bestTrack()->dz(PVpos)) > 17) continue;
+	nchdaugthers++;
+      }
+    }    
+    //
+    //Fill the pt ratio and pt rel
+    //
+    //No jets found
+    if(dR > dRmax_){
+      ptratio.push_back(-99);
+      ptrel.push_back(-99);
+      btagcsv.push_back(-999.);
+      nchargeddaughers.push_back(-1);
+    }
+    else{
+      if ((jet-mu).Rho()<0.0001) 
+	jet=mu; 
+      else {
+	jet -= mu/jecL1;
+	jet *= jecL1L2L3Res;
+	jet += mu;
+      }
+      ptratio.push_back(mu.pt()/jet.pt());
+      TLorentzVector tmp_mu, tmp_jet;
+      tmp_mu.SetPxPyPzE(mu.px(),mu.py(),mu.pz(),mu.E());
+      tmp_jet.SetPxPyPzE(jet.px(),jet.py(),jet.pz(),jet.E());
+      if ((tmp_jet-tmp_mu).Rho()<0.0001)  ptrel.push_back(0.);
+      else 	                          ptrel.push_back(tmp_mu.Perp((tmp_jet-tmp_mu).Vect()));
+      btagcsv.push_back(bjet);
+      nchargeddaughers.push_back(nchdaugthers);
+    }
+
+    //    printf ("muon: %4.2f %4.2f %4.2f %4.2f %4.2f %4.2f %4.2f\n", mu.pt(),mu.eta(),mu.phi(), ptrel.back(), ptratio.back(), nchargeddaughers.back(), btagcsv.back());
+
+
+  }//end muon loop
+  
+  
+  /// Filling variables previously computed
```

Comment:
Maybe you could do a method as [here](https://github.com/cms-analysis/MuonAnalysis-TagAndProbe/blob/b900acf4ae2114d83e309213a503abae8b22a507/plugins/MuonMetMT.cc#L82) for example to avoid repeating the filler 4 times